common/unsafe/src/main/java/org/apache/spark/sql/catalyst/util/CollationAwareUTF8String.java

```diff
@@ -339,10 +339,11 @@ public static UTF8String lowercaseReplace(final UTF8String src, final UTF8String
    * @return the uppercase string
    */
   public static UTF8String toUpperCase(final UTF8String target) {
-    return UTF8String.fromString(toUpperCase(target.toString()));
+    if (target.isFullAscii()) return target.toUpperCaseAscii();
+    return UTF8String.fromString(toUpperCaseSlow(target.toString()));
   }
 
-  public static String toUpperCase(final String target) {
+  private static String toUpperCaseSlow(final String target) {
```

Comment:
Why don't we make this method also take and return `UTF8String`? I think it would be cleaner this way. How we call ICU becomes something completely internal to this method.

Reply:
yeah, I guess we could do this - fixing now